drivers/flash/flash_ambiq.c

```diff
@@ -0,0 +1,182 @@
+
+#include <am_mcu_apollo.h>
+
+LOG_MODULE_REGISTER(flash_ambiq, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE      DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_ADDR      DT_REG_ADDR(SOC_NV_FLASH_NODE)
+#define SOC_NV_FLASH_SIZE      DT_REG_SIZE(SOC_NV_FLASH_NODE)
+#define FLASH_WRITE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, write_block_size)
+#define FLASH_ERASE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+#if defined(CONFIG_MULTITHREADING)
+static struct k_sem flash_ambiq_sem;
+#define FLASH_SEM_INIT() k_sem_init(&flash_ambiq_sem, 1, 1)
+#define FLASH_SEM_TAKE() k_sem_take(&flash_ambiq_sem, K_FOREVER)
+#define FLASH_SEM_GIVE() k_sem_give(&flash_ambiq_sem)
+#else
+#define FLASH_SEM_INIT()
+#define FLASH_SEM_TAKE()
+#define FLASH_SEM_GIVE()
+#endif /* CONFIG_MULTITHREADING */
+
+static const struct flash_parameters flash_ambiq_parameters = {
+	.write_block_size = FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+static bool flash_ambiq_valid_range(off_t offset, size_t len)
+{
+	if ((offset < 0) || (offset + len) > SOC_NV_FLASH_SIZE) {
+		return false;
+	}
+
+	return true;
+}
+
+static int flash_ambiq_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	memcpy(data, (uint8_t *)(SOC_NV_FLASH_ADDR + offset), len);
+
+	return 0;
+}
+
+static int flash_ambiq_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	int ret = 0;
+	uint32_t critical = 0;
+	uint32_t aligned[4] = {0, 0, 0, 0};
+	uint32_t *src = (uint32_t *)data;
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* The destination address must be 16 bytes aligned and the write bytes must be a multiple
+	 * of 16, which will be checked in HAL.
+	 */
+
+	if (len == 0) {
+		return 0;
+	}
+
+	FLASH_SEM_TAKE();
+
+	critical = am_hal_interrupt_master_disable();
+	for (int i = 0; i < len / 16; i++) {
+		for (int j = 0; j < 4; j++) {
+			/* Make sure the source data is 4-byte aligned. */
+			aligned[j] = UNALIGNED_GET((uint32_t *)src);
+			src++;
+		}
+		ret = am_hal_mram_main_program(AM_HAL_MRAM_PROGRAM_KEY, aligned,
+					       (uint32_t *)(SOC_NV_FLASH_ADDR + offset + i * 16),
+					       4);
+		if (ret) {
+			break;
+		}
+	}
+	am_hal_interrupt_master_set(critical);
+
+	FLASH_SEM_GIVE();
+
+	return ret;
+}
+
+static int flash_ambiq_erase(const struct device *dev, off_t offset, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	int ret = 0;
+	uint32_t erase_word = 0;
+
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	FLASH_SEM_TAKE();
+
+	erase_word = ((uint32_t)(flash_ambiq_parameters.erase_value << 24)) |
+		     ((uint32_t)(flash_ambiq_parameters.erase_value << 16)) |
+		     ((uint32_t)(flash_ambiq_parameters.erase_value << 8)) |
+		     ((uint32_t)flash_ambiq_parameters.erase_value);
```

Comment:
Thad does not need to be within SEM.
Also, unless you are supporting multiple instances with different erase_values, the 32bit erase value may be pre-calculated for this module at compile time.